scintillometry/utils/phase_utils.py

```diff
@@ -0,0 +1,135 @@
+# Licensed under the GPLv3 - see LICENSE
+
+import warnings
+
+import numpy as np
+import astropy.units as u
+import astropy.time as time
+
+try:
+    from .pint_utils import PintToas, make_toa_list
+    from pint.models import get_model
+except ImportError:
+    pass
+from .predictor import Polyco
+
+
+__all__ = ['PintPhase', 'PolycoPhase']
+
+
+class PintPhase(object):
+    """A utility class designated for computing phase using PINT.
+       Parameter
+       ---------
+       par_file : str
+           TEMPO/TEMPO2 style parameter fileself.
+       obs : str
+           Observatory name or observatory codeself.
+       obs_freq : float or `~astropy.units.Quantity`.
+           Observing frequency default units is 'MHz'
+       ephem : str
+           The solar system ephemeris version.
+       bipm_version : str
+           The BIPM clock correction version.
+
+       Note
+       ----
+       This method provides high precision phase calculation. The fractional
+       phase is between -0.5 to 0.5.
+    """
+    def __init__(self, par_file, obs, obs_freq, ephem='de436', include_bipm=True,
+                 bipm_version='BIPM2015', include_gps=True, planets=False,
+                 tdb_method="default", control_from_par=False, **kwargs):
+        self.par_file = par_file
+        self.model = get_model(self.par_file)
+        self.toa_maker = PintToas(ephem=ephem, include_bipm=include_bipm,
+                                  bipm_version=bipm_version,
+                                  include_gps=include_gps,
+                                  planets=planets, tdb_method=tdb_method,
+                                  **kwargs)
+        self.obs = obs
+        self.obs_freq = obs_freq
+
+    def __call__(self, t):
+        """Compute the apparent phase at one or more time stemp
+
+           Parameter
+           ---------
+           t : `astropy.time.Time` object
+               The input time stemps.
+
+           Return
+           ------
+           The apparent pulse phase at time t. The phases are return as
+           'integer phase, fractional phase'
+        """
+        toas = self._make_toas(t)
+        ph = self.model.phase(toas)
+        return u.Quantity((ph.int, ph.frac), u.cycle)
```

Comment:
I think it actually makes more sense to return two quantities, which has the advantage that a copy can be avoided, i.e.,
```
return u.Quantity(ph.int, u.cycle, copy=False), u.Quantity(ph.frac, u.cycle, copy=False)
```